### DISCUSSION TITLE
[How to] Minimal example compilation and run not working ld: Undefined symbols

### DISCUSSION THREAD

**Discussion by mfouesneau** [alandefreitas/matplotplusplus#96] [Q&A]
I'm trying to run a very simple example taken from the examples directory.

```c++
#include <cmath>
#include <matplot/matplot.h>

int main() {
    using namespace matplot;
    std::vector<double> x = matplot::linspace(0, 2 * pi);
    std::vector<double> y = matplot::transform(x, [](auto x) { return sin(x); });

    matplot::plot(x, y, "-o");
    matplot::hold(on);
    matplot::plot(x, transform(y, [](auto y) { return -y; }), "--xr");
    matplot::plot(x, transform(x, [](auto x) { return x / pi - 1.; }), "-:gs");
    matplot::plot({1.0, 0.7, 0.4, 0.0, -0.4, -0.7, -1}, "k");

    matplot::show();
    return 0;
}
```

I installed matplot++ with Brew 

My CMake configuration is the following
```
cmake_minimum_required(VERSION 3.14.0)
project(matplot_example VERSION 0.1.0)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

# set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -O2 ")
# set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -Wall -O2 ")

find_package(Matplot++ REQUIRED)

if (Matplot++_FOUND)
    add_executable(example example.cpp)
    target_compile_features(example PRIVATE cxx_std_17)
    target_link_libraries(example PUBLIC Matplot++::matplot)
    install(TARGETS example DESTINATION bin)
endif()
```

It seems to find all libraries etc, but when creating an executable I get ld: Undefined symbols errors.

Any idea what's going wrong??

**Comment by alandefreitas** [comment]
It should be OK. Can you post the error messages you get? You can also try to install it a different way just to make sure it's not the brew formula. Although the formula seems ok.

**Reply by mfouesneau** [reply]
```
[proc] Executing command: /usr/local/bin/cmake --build ./build --config Debug --target all -- -j 14
[build] [1/2  50% :: 2.128] Building CXX object CMakeFiles/example.dir/example.cpp.o
[build] [2/2 100% :: 2.209] Linking CXX executable example
[build] FAILED: example 
[build] : && /usr/bin/clang++ -g -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/example.dir/example.cpp.o -o example  /usr/local/lib/libmatplot.dylib && :
[build] Undefined symbols for architecture x86_64:
[build]   "matplot::axes_type::plot(std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::basic_string_view<char, std::__1::char_traits<char> >)", referenced from:
[build]       matplot::plot(std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::basic_string_view<char, std::__1::char_traits<char> >) in example.cpp.o
[build]   "matplot::axes_type::plot(std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::basic_string_view<char, std::__1::char_traits<char> >)", referenced from:
[build]       auto matplot::plot<std::__1::vector<double, std::__1::allocator<double> >, std::__1::vector<double, std::__1::allocator<double> >&, char const (&) [3]>(std::__1::vector<double, std::__1::allocator<double> >, std::__1::vector<double, std::__1::allocator<double> >&&&, char const (&&&) [3]) in example.cpp.o
[build]       auto matplot::plot<std::__1::vector<double, std::__1::allocator<double> >, std::__1::vector<double, std::__1::allocator<double> >, char const (&) [5]>(std::__1::vector<double, std::__1::allocator<double> >, std::__1::vector<double, std::__1::allocator<double> >&&, char const (&&&) [5]) in example.cpp.o
[build]   "matplot::transform(std::__1::vector<double, std::__1::allocator<double> > const&, std::__1::function<double (double)>)", referenced from:
[build]       _main in example.cpp.o
[build] ld: symbol(s) not found for architecture x86_64
[build] clang: error: linker command failed with exit code 1 (use -v to see invocation)
[build] ninja: build stopped: subcommand failed.
[cmakefileapi-parser] Code model version (2.1) of cmake-file-api is unexpected. Expecting (2.0). IntelliSense configuration may be incorrect.
[cmakefileapi-parser] Code model version (2.1) of cmake-file-api is unexpected. Expecting (2.0). IntelliSense configuration may be incorrect.
[build] Build finished with exit code 1
```

**Reply by alandefreitas** [reply]
Everything seems to be OK with your code. You can try to install it with the official packages or build from source. If it works, maybe report an error with the brew package. We incentivize packages but we don't support them.

**Reply by mfouesneau** [reply]
Compiling from source fails
```
-- A library with LAPACK API found.
CMake Error at cmake/FindFilesystem.cmake:246 (message):
  Cannot run simple program using std::filesystem
Call Stack (most recent call first):
  source/matplot/CMakeLists.txt:1 (find_package)


-- Configuring incomplete, errors occurred!
See also "CMakeFiles/CMakeOutput.log".
See also "CMakeFiles/CMakeError.log".
```

**Reply by alandefreitas** [reply]
Mmmm... Your C++ compiler is too old then:

```
Cannot run simple program using std::filesystem
```

**Comment by mfouesneau** [comment]
How do I update it? xcode 11.3 is the most of to date for my computer. (I cannot upgrade the system to obtain more recent version)

**Reply by alandefreitas** [reply]
There are some instructions here that might help.
https://alandefreitas.github.io/matplotplusplus/integration/build-from-source/dependencies.html

You can also try to compile any snippet of code with the [filesystem library](https://en.cppreference.com/w/cpp/filesystem) to make sure things are working.

**Comment by mfouesneau** [comment]
Clang 11.0.0
cmake 3.18.4

I cannot compile the source. Same as above.

**Reply by alandefreitas** [reply]
Have you tried using the C++17 at all regardless of matplot++? Try compiling something like this:

```cpp
#include <iostream>
#include <filesystem>
namespace fs = std::filesystem;
int main()
{
    fs::path p1 = "/usr/lib/sendmail.cf"; // portable format
    fs::path p2 = "C:\\users\\abcdef\\AppData\\Local\\Temp\\"; // native format
    fs::path p3 = L"D:/猫.txt"; // wide string
 
    std::cout << "p1 = " << p1 << '\n'
              << "p2 = " << p2 << '\n'
              << "p3 = " << p3 << '\n';
}
```

**Comment by mfouesneau** [comment]
'path' has been explicitly marked unavailable here

but filesystem seems to be found...

**Reply by alandefreitas** [reply]
Well... you have to update to C++17 somehow. Or maybe just set it up. It seems like you're using a C++14 version where they were beginning to implement C++17 in the experimental headers.

`path` is probably the most basic features of the filesystem library. The complete filesystem library has been available in clang [for a long long time](https://clang.llvm.org/cxx_status.html).

Macs and Mac OS have been coming with C++17 by default for a while. At most, if the Mac OS is not recent enough, some people need to update XCode in the App Store. I never seen it take more than that.

**Comment by mfouesneau** [comment]
apparently not in 10.14... : ```'path' is unavailable: introduced in macOS 10.15```. C++17 standard is understood but xcode is locked on 10.14 to 11.3. Do you know how to install a newer version?

**Reply by alandefreitas** [reply]
Yes. C++17 was not implemented yet. They just had C++14 and a few experimental C++17 headers at this point.

I'm on Mac OS 10.15.6 here. If updating from the XCode App Store doesn't solve the problem, then I think the only option is to directly follow the instructions to update clang from https://clang.llvm.org. 

This is a little more complicated because I think Apple wants to incentivize (to force) people to update XCode from the App Store and to update Mac OS to the latest version in general. 

I don't know if the instructions there are easy to follow, but I've seen some answers on stack overflow about that and there might be a brew formula for that.

**Reply by mfouesneau** [reply]
I could not update my system or it does not update fully on 10.14.6.

**Reply by mfouesneau** [reply]
XCode is stuck on older versions of MacOS, which does not help. I could not even install CML updates without some hacking.

**Comment by mfouesneau** [comment]
I am trying with gcc 10  

(If anyone reading these lines has an error -lSystem not found on a mac, reinstall gcc)

**Reply by alandefreitas** [reply]
In my experience, if you're working on Mac OS, installing GCC (and the corresponding std library) is trickier than updating clang and the corresponding std library, which is already sort of intricate. I think you would need manually link the std library in cmake or something like that to make it work. I don't really remember the process by heart but I remember it convinced me to only use clang+Mac Os, GCC+Linux, and MSVC+Windows.

**Reply by mfouesneau** [reply]
I fully agree that mixing is not optimal. I would also prefer using clang+Mac Os, GCC+Linux, and MSVC+Windows to avoid headaches.
Brew does pretty well at updating/installing gcc, but I would also like to have something simpler.

**Reply by alandefreitas** [reply]
Have you tried building [clang](https://clang.llvm.org/get_started.html) and [libc++](https://libcxx.llvm.org/docs/BuildingLibcxx.html) from source? Apple doesn't really incentivize people to do that, but there's nothing they can do if you build from source. Just don't replace the system compiler to be safe. 

If their documentation doesn't fail, it should be something like:

```
git clone https://github.com/llvm/llvm-project.git
cd llvm-project
mkdir build (in-tree build is not supported)
cd build
cmake -DLLVM_ENABLE_PROJECTS=clang -G "Unix Makefiles" ../llvm
make
```

and 

```
git clone https://github.com/llvm/llvm-project.git
cd llvm-project
mkdir build && cd build
cmake -DCMAKE_C_COMPILER=clang \
        -DCMAKE_CXX_COMPILER=clang++ \
        -DLLVM_ENABLE_PROJECTS="libcxx;libcxxabi" \
        ../llvm
make # Build
make check-cxx # Test
make install-cxx install-cxxabi # Install
```

And then tell cmake to explicitly use this new compiler and this new stdlib so it compiles things with something like:

```
<clang-install-prefix>/clang++ -std=c++17 -stdlib=libc++ -nostdinc++ \
          -I<libcxx-install-prefix>/include/c++/v1 \
          -L<libcxx-install-prefix>/lib \
          -Wl,-rpath,<libcxx-install-prefix>/lib \
          test.cpp
```
